Mathlib/Tactic/Rewrites.lean

```diff
@@ -118,15 +118,19 @@ structure RewriteResult where
   /-- Can the new goal in `result` be closed by `with_reducible rfl`? -/
   -- This is an `Option` so that it can be computed lazily.
   rfl? : Option Bool
+  /-- The metavariable context after the rewrite.
+  This needs to be stored as part of the result so we can backtrack the state. -/
+  mctx : MetavarContext
 
 /-- Update a `RewriteResult` by filling in the `rfl?` field if it is currently `none`,
 to reflect whether the remaining goal can be closed by `with_reducible rfl`. -/
 def RewriteResult.computeRfl (r : RewriteResult) : MetaM RewriteResult := do
   if let some _ := r.rfl? then
     return r
   try
-    (← mkFreshExprMVar r.result.eNew).mvarId!.rfl
-    pure { r with rfl? := some true }
+    withoutModifyingState <| withMCtx r.mctx do
+      (← mkFreshExprMVar r.result.eNew).mvarId!.rfl
+      pure { r with rfl? := some true, mctx := ← getMCtx }
```

Comment:
```suggestion
      pure { r with rfl? := some true }
```
I think we should not update the `mctx` to the post-`rfl` mctx here. I thought about this carefully, and would be happy to expand on it if you'd be interested—but to keep it concise, three facts might be sufficient to show that we shouldn't be doing this:
* this `mctx` remembers the temporary `mvar` that we just made with `mkFreshExprMVar`, which shouldn't exist outside of this "`rfl` simulation".
* when we're feeling lucky, we run `try rfl` *after* setting the mctx. This means we ought to start with an `MCtx` where `rfl` hasn't yet acted.
* likewise, we don't also update `r.result.eNew` or `r.result.eProof` here, and those should stay in lockstep with the `mctx`.

Reply:
Thanks, I agree.